src/Integration.Vsix/Suppression/LiveIssueFactory.cs

```diff
@@ -0,0 +1,137 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.VisualStudio.Shell.Interop;
+
+/* To map from a diagnostic to a SonarQube issue we need to work out the SQ moduleId
+ * corresponding to the MSBuild project.
+ * 
+ * To speed things up, we cache the mapping from project file path to project id
+ * for the loaded projects (which means we need to rebuild the mapping when the
+ * solution changes).
+ * 
+ * Note: this class does not monitor changes to the solution. If the solution/binding
+ * changes then create a new instance.
+ */
+
+namespace SonarLint.VisualStudio.Integration.Vsix.Suppression
+{
+    /// <summary>
+    /// Factory for <see cref="LiveIssue"/>s i.e. diagnostics that are decorated with
+    /// additional information required to map to issues in the SonarQube server format
+    /// </summary>
+    internal sealed class LiveIssueFactory
+    {
+        private readonly Workspace workspace;
+        private readonly IServiceProvider serviceProvider;
+
+        /// <summary>
+        /// Mapping from full project file path to the unique project id
+        /// </summary>
+        private Dictionary<string, string> projectPathToProjectIdMap;
+
+        public LiveIssueFactory(IServiceProvider serviceProvider)
+        {
+            this.serviceProvider = serviceProvider;
+
+            var componentModel = (IComponentModel)serviceProvider.GetService(typeof(SComponentModel));
+            workspace = componentModel.GetService<VisualStudioWorkspace>();
+
+            BuildProjectPathToIdMap();
+        }
+
+        /// <summary>
+        /// Attempts to fetch the extra information required to map from a Roslyn issue
+        /// to a SonarQube server issue. Returns null if there is not enough information
+        /// to produce the mapping.
+        /// </summary>
+        public LiveIssue TryCreate(Diagnostic diagnostic)
```

Comment:
I would rename the method to `Create` because when I see `TryXXX` I always expect a bool return and a out parameter (not a big fan of this syntax by the way).